src/opt/Optimizer/TypedDesugarer.hs

```diff
@@ -0,0 +1,257 @@
+
+-- Modular dependancies
+import Identifiers
+import AST.AST
+import AST.Util
+import Identifiers
+import qualified AST.Meta as Meta
+import Types
+import Typechecker.Environment
+import Typechecker.TypeError
+import AST.PrettyPrinter
+
+desugarTypedProgram :: Program -> Program
+desugarTypedProgram p@(Program{classes, traits, functions}) =
+    p{classes = map desugarClass classes
+     ,traits = map desugarTrait traits
+     ,functions = map desugarFunction functions}
+    where
+      desugarFunction f@(Function{funbody}) =
+          f{funbody = desugarExpr funbody}
+
+      desugarTrait t@(Trait{tmethods}) =
+          t{tmethods = map desugarMethod tmethods}
+
+      desugarClass c@(Class{cname, cmethods})=
+          c{cmethods = map desugarMethod cmethods}
+
+      desugarMethod m =
+          m{mbody = desugarExpr (mbody m)}
+
+      desugarExpr ast =
+          foldl (\ast opt -> opt ast) ast desugarPasses
+
+-- | The functions in this list will be performed in order during desugaring
+desugarPasses :: [Expr -> Expr]
+desugarPasses = [desugarAndBoxForInSeq, desugarAndBoxForNotInSeq]
+
+desugarAndBoxForInSeq = extend desugarAndBoxForInSeq'
+  where desugarAndBoxForInSeq' e@Seq{} = boxed e
+        desugarAndBoxForInSeq' e = e
+        boxed e = boxForInSeq e
+
+desugarAndBoxForNotInSeq = extend desugarAndBoxForNotInSeq'
+  where desugarAndBoxForNotInSeq' for@For{} = forBoxed for []
+        desugarAndBoxForNotInSeq' e = e
+
+
+
+-- Desugars a for-loop into nested calls to map and flatMap and foreach:
+--
+-- for x <- listA, y <- listB, z <- ListC do
+--      fun
+-- end
+--
+-- into listA.flatMap(listB.flatMap(listC.map(fun)))
+forDesugared :: Expr -> Expr
+forDesugared e@For{emeta, sources, body} =
+  let n = length sources
+      collectionType = getType $ collection $ head sources
+      callNameList = if (not (AST.AST.isCaptured e)) || (unitType == getType body) || (isRangeObjectType collectionType)
+                     then replicate n (Name "foreach")
+                     else replicate (n-1) (Name "flatMap") ++ [Name "map"]
+      revSources = reverse sources
+      elemType = getType body
+      desugaredFor = nestCalls emeta callNameList sources body elemType
+  in  desugaredFor
+forDesugared m = m
+
+nestCalls :: Meta.Meta Expr -> [Name] -> [ForSource] -> Expr -> Type -> Expr
+nestCalls meta (name:_) (fs:[]) body elemType = intoCall meta name fs body elemType
+nestCalls meta (name:restOfNames) (fs:restFS) body elemType =
+  let nestedCall = intoCall meta name fs body elemType
+  in nestCalls meta restOfNames restFS nestedCall elemType
+
+intoCall :: Meta.Meta Expr -> Name -> ForSource -> Expr -> Type -> Expr
+intoCall met callName ForSource{fsName, fsTy, collection} bodyOrMethodCall elemType =
+  if isRefType (getType collection)
+  then let param = [intoParam met Val fsName fsTy]
+           arguments = [intoClosure met param Nothing bodyOrMethodCall]
+           elemT = if callName == Name "foreach" -- this feels iffy
```

Comment:
since you have written a bunch of time `Name "foreach"`, maybe it makes sense to define it and create a helper function, `isForeach e`, e.g.
```haskell
foreachName = Name "foreach"
isForeach e = e == foreachName
```

Reply:
my feeling is that these kind of constants are written somewhere in the code base